config.yml

```diff
@@ -1,98 +1,121 @@
 
-common_settings:
-  # common settings for ALL projects in ALL groups
-
-  # keys and values here are as described at https://docs.gitlab.com/ee/api/project_level_variables.html#create-variable
-  secret_variables:
-    # this name is not actually used, it's just for you
-    a_secret_you_want_to_add_to_all_projects_in_your_gitlab_instance:
-      key: A_VERY_COMMONLY_USED_PASSWORD
-      value: "ThisIsAVerySecretPassword"
-      delete: false  # Set to true to delete this variable
-
-  # keys and values here are as described at https://docs.gitlab.com/ee/api/group_level_variables.html#create-variable
-  group_secret_variables:
-    # this name is not actually used, it's just for you
-    a_secret_you_want_to_add_to_all_groups_in_your_gitlab_instance:
-      key: A_VERY_COMMONLY_USED_PASSWORD
-      value: "ThisIsAVerySecretPassword"
-      variable_type: env_var #or file
-      protected: false #
-      delete: false  # Set to true to delete this variable
-
-group_settings:
-  # settings for ALL projects in 'my-group' group
-  "my-group":
-    deploy_keys:
-      # this name is not actually used, it's just for you
-      a_friendly_deploy_key_name:
-        # you have to pass whole SSH key content here even if GitLab already has this key added and you just assign it
-        # to another project here
-        # this is a limitation of GitLab API
-        key: "ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAACAQDB2QKx6BPzL (...)"
-        title: ssh_key_name_that_is_shown_in_gitlab
-        # note that you can set this to `true` or `false` only on the first assignment of the key / its creation
-        # see https://gitlab.com/gitlab-org/gitlab-ce/issues/30021#note_39567845
-        # this is a limitation of GitLab API
-        can_push: false
-      # this name is not actually used, it's just for you
-      another_friendly_deploy_key_name:
-        key: "ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAACAQDtbyEK66RXg (...)"
-        title: other_ssh_key_name_that_is_shown_in_gitlab
-        can_push: false
 
-    # keys and values here are as described at https://docs.gitlab.com/ee/api/project_level_variables.html#create-variable
-    secret_variables:
-      # this name is not actually used, it's just for you
-      a_friendly_secret_variable_name:
-        key: SSH_PRIVATE_KEY_BASE64
-        value: "LS0tLS1CRUdJTiBSU0EgUFJJVkFURSBLRVktLS0tLQpNSUl (...)"
+# The main config starts with this key
+projects_and_groups:
+
+  # On this indentation level you choose what groups and/or projects you want to configure.
+  #
+  # Here are examples of possible dict key.
+  #
+  # This matches all the groups AND all the projects EXCEPT personal projects!
+  # "*":
+  #
+  # This matches all the personal projects.
+  # "<all-users>":
+  #
+  # This matches the "my-group" group AND projects in it.
+  # my-group:
+  #
+  # This matches the "my-subgroup" subgroup within the "my-group" group AND projects in it.
```

Comment:
will this be recursive? would this also affect `my-group/my-subgroup/my-sub-subgroup`?

Reply:
That's a good question. What would be expected behaviour for you? I am currently thinking about few possible approaches:
1. Keep is as it is - no recursion. This means though that all subgroups would have to be explicitly specified (and it weakens the power of pattern matching that we would like to introduce here).
2. Make it non-recursive by default and allow recursive search to be enabled (either only on single pattern level, globally, or both).
3. Make it always recursive (with option to disable it? not sure if it makes sense though).

Reply:
Another option - a `recursive: true/false` property to be defined on each level.

This would allow to e.g. include only a subgroup with sub-subgroups without having to resort to a *list* of patterns.

Argh, too many options :)